numpy/lib/function_base.py

```diff
@@ -2669,6 +2689,26 @@ class vectorize(object):
            <http://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html>`_.
     """
 
+    # __new__ is overriden as a clean way to provide decorator-with-keywords
+    # syntax while allowing 'vectorize' to remain as the type of vectorized
+    # functions.
+    def __new__(cls, pyfunc=None, *args, **kwargs):
+
+        def vectorize_decorator(pyfunc):
+            """ ``vectorize`` with presupplied keyword arguments. """
+            # prevent an edge case where None would produce another decorator
+            if not hasattr(pyfunc, '__call__'):
+                raise TypeError('the wrapped object must be callable')
+            return cls(pyfunc, *args, **kwargs)
+
+        # For backwards compatibility, subclasses must not use the overloaded
```

Comment:
Readable is always in the eyes of the beholder: I find it visually distracting to have a function defined and then it not being used in the by far most common case. How about just:
```
if cls is vectorize and pyfunc is None:
    def vectorize_decorator(...):
        ...
    return vectorize decorator
else:
    return object.__new__(cls)
```

Reply:
I don't care too much, but I like the following, with early escapes:

```python
if cls is not vectorize:
    # Overloading subclass behaviour would make assumptions about __init__
    return object.__new__(cls)

if pyfunc is not None:
    # decorator not needed - construct as normal
    return object.__new__(cls)


def vectorize_decorator(...):
    ...
return vectorize decorator
```

Reply:
Honestly (and I know this is beating a dead horse at this point :stuck_out_tongue:) but I see `vectorize_decorator` is the early escape.

In fact, knowing me, I probably initially wrote it as:

```
if cls is vectorize and pyfunc is None:
    return vectorize_decorator

return object.__new__(cls)
```

Then I probably went back and added the `else:` after browsing around and finding [existing code like this](https://github.com/ExpHP/numpy/blob/8c57eb11d05fc2dc05d8f6754ede3e1e7bae7e0e/numpy/lib/function_base.py#L1822-L1825)

Reply:
Actually, why not expose `vectorize.decorator` as a public `@classmethod`? That way derived classes can easily opt-in

Reply:
That sounds like a good idea. I will need to play around with the idea some in order to think of good recommendations for how a subclass can opt in (and when it may not be suitable)

Reply:
@eric-wieser Now that I'm trying it, I don't see a straightforward path to exposing it, since `vectorize_decorator` is closed over `args` and `kwargs` from the surrounding context.  Was there something particular you had in mind?